### PR TITLE
Path forward for removing DT_FLASH_DEV_NAME

### PR DESCRIPTION
**Pull request by mbolivar-nordic**
Based on https://github.com/zephyrproject-rtos/zephyr/pull/23890, so ignore the first two commits.

This PR adds devicetree accessors for zephyr-specific properties of the /chosen node. The only user added is a DT_ZEPHYR_FLASH node identifier for the zephyr,flash chosen property.

Adding this allows replacing users of the DT_FLASH_DEV_NAME macro defined in dts_fixup.h files with a more generic API that just gets the label property value of the DT_ZEPHYR_FLASH node, if it exists. Backwards compatibility is preserved by falling back on DT_FLASH_DEV_NAME if the chosen zephyr,flash node is not available.

This plus a couple of other changes were enough to get the flash_shell sample building on various nRF boards, and lightly tested on nRF51-DK, as proof of concept that it allows removing DT_FLASH_DEV_NAME from our SoC dts_fixup.h files.

If this looks good, other vendors can follow suit, and we can remove DT_FLASH_DEV_NAME entirely.

### CODE REVIEW

File: include/devicetree/zephyr.h [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+
+/*
+ * This file is currently deliberately not defining macros for some
+ * existing zephyr,foo chosen nodes, such as zephyr,sram, until there
+ * are users for them. Feel free to extend it as needed.
+ */
+
+#if DT_HAS_CHOSEN(zephyr_flash)
+/** Node identifier for the /chosen node's zephyr,flash property value. */
+#define DT_ZEPHYR_FLASH DT_CHOSEN(zephyr_flash)
```
Comment (galak) [inline include/devicetree/zephyr.h]: Feel like putting CHOSEN in the define name.. so either `DT_CHOSEN_ZEPHYR_FLASH` (which would map to the path) or `DT_ZEPHYR_CHOSEN_FLASH`

File: include/devicetree/zephyr.h [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+/** The /chosen zephyr,flash node's label property. */
+#define DT_ZEPHYR_FLASH_LABEL DT_LABEL(DT_ZEPHYR_FLASH)
+#elif defined(DT_FLASH_DEV_NAME) /* fall back on dts_fixup.h */
+#define DT_ZEPHYR_FLASH_LABEL DT_FLASH_DEV_NAME
```
Comment (galak) [inline include/devicetree/zephyr.h]: I'd stick a __deprecated on this for now, we can remove right before release, but that will catch any PRs or in-tree usage.

File: samples/bluetooth/mesh_demo/src/microbit.c [hunk 3]
```diff
@@ -249,7 +249,7 @@ void board_init(u16_t *addr)
 {
 	struct mb_display *disp = mb_display_get();
 
-	nvm = device_get_binding(DT_FLASH_DEV_NAME);
+	nvm = device_get_binding(DT_ZEPHYR_FLASH_LABEL);
```
Comment (hakonfam) [inline samples/bluetooth/mesh_demo/src/microbit.c]: Just curious, why "most" and not "all" ?

**Review by hakonfam — APPROVED** [review]
Not a DT expert, but my shallow review finds this to be a great improvement :)

**Review by nvlsianpu — COMMENTED** [review]
patch looks good to me,  just like idea of having chosen in new labels name.

File: include/devicetree/zephyr.h (hunk 2 above)
Comment (mbolivar-nordic) [reply include/devicetree/zephyr.h:32]: > I'd stick a __deprecated on this for now, we can remove right before release, but that will catch any PRs or in-tree usage.

I think you and @nashif ought to talk about this. From what I understood on Slack, he believes devicetree ought to be optional as a long-term thing. I'm confused about what's going on.

Comment (galak) [reply include/devicetree/zephyr.h:32]: Ok, not sure how that is related.

Comment (mbolivar-nordic) [reply include/devicetree/zephyr.h:32]: How can we remove this if we always need to support non-DT ways to name this string?

Comment (mbolivar-nordic) [reply include/devicetree/zephyr.h:32]: I.e. this is a specific instance of a more general situation where I am not sure if we're going to deprecate and remove non-devicetree ways of naming and creating devices, or if we need fallbacks and fixups forever.

**Comment by mbolivar-nordic** [comment]
@andrewboie @nvlsianpu 

I'm trying to understand what's going on with https://github.com/zephyrproject-rtos/zephyr/pull/17349/commits/878dfc3d2daca633050da9744f33250a8d05568a as it affects this PR.

It looks like an inconsistent change, because DT_FLASH_DEV_NAME is still DT_ZEPHYR_SIM_FLASH_SIM_FLASH_LABEL, even though the chosen node is moving from flash_sim0 to flash0.

I don't understand the commit message either, because that says it fixes #15832, but that's a pull request, not an issue, and you can't fix a pull request.

Can anyone clear up how these flash nodes are used in qemu_x86 and suggest a consistent choice for the zephyr,flash chosen node? @andrewboie ?

**Comment by mbolivar-nordic** [comment]
> I'm trying to understand what's going on with [878dfc3](https://github.com/zephyrproject-rtos/zephyr/commit/878dfc3d2daca633050da9744f33250a8d05568a) as it affects this PR.

I think I'm just going to ignore inconsistencies like this with a bit of extra work on the backwards compatibility. This affects at least the following boards:

- altera_max10
- native_posix
- qemu_x86 (and friends)

File: include/devicetree/zephyr.h (hunk 1 above)
Comment (mbolivar-nordic) [reply include/devicetree/zephyr.h]: DT_CHOSEN_ZEPHYR_FLASH isn't any shorter than DT_CHOSEN(zephyr_flash) and neither is really needed, so I've removed this macro and am focusing on the label.

**Review by nvlsianpu — APPROVED** [review]

**Review by erwango — APPROVED** [review]

**Comment by mbolivar-nordic** [comment]
Further investigation by @galak made it clear that this will require bigger changes than originally anticipated. Closing for now; we will get devicetree/zephyr.h merged via https://github.com/zephyrproject-rtos/zephyr/pull/24127 instead. Thanks for everyone's review and discussion today at dev-review.

**Pull request closed without merging**